Refactor/refactor navigation

Joseph-Sameh-0 opened the pull request:
This pull request introduces significant changes to the structure and functionality of the project, including module restructuring, dependency updates, and the addition of a new navigation module. The most important changes center around refactoring the application structure, updating dependencies, and introducing a modular navigation system.

### Application Refactoring and Renaming:
* The `SearchApp` class was renamed to `AflamiApplication`, and its package was updated to `com.paris_2.aflami`. The application now initializes Koin with updated modules, including a newly added `FeatureAPIModule`. (`app/src/main/AndroidManifest.xml`, `app/src/main/java/com/paris_2/aflami/AflamiApplication.kt`) [[1]](diffhunk://#diff-7fa6aef292187a049f7a4d6060d8df3ba212d838789c78940bd363344b1c38cdR9) [[2]](diffhunk://#diff-1a61c6610f478d65d07a28f90d58632bf6c654e0b89f8c7ef6f0c32578cf4052L1-R41)

* Multiple files were renamed and moved from their original locations (e.g., `feature/search/searchUi` to `app/src/main/java/com/paris_2/aflami/di`) to align with the new application structure. (`DataSourceModule.kt`, `NetworkModule.kt`, `RoomModule.kt`, etc.) [[1]](diffhunk://#diff-0dc9a413099cbd33d4a30ecdda8166ddf6882b01e2f760e1c6a0fa742ac73674L1-R1) [[2]](diffhunk://#diff-64d43872aeda7eca1074dd279cf378cfe14d16c397f6ba402ffd4db74203dba0L1-R1) [[3]](diffhunk://#diff-027b207a979bb620a93b6ddc793eb5a66e0bd923638e9e2968733ff8fe7af64dL1-R1)

### Dependency Updates:
* Added new dependencies to `app/build.gradle.kts`, including `koin.workmanager`, `work.runtime.ktx`, and Room database dependencies (`room.runtime` and `room.compiler` with KSP). Several project modules were also added as dependencies. (`app/build.gradle.kts`)

* Introduced the `ksp` plugin to the project for annotation processing. (`app/build.gradle.kts`)

### Navigation System:
* Added a new `appNavigation` module with its own `build.gradle.kts` and ProGuard rules. This module includes dependencies for Jetpack Compose Navigation and Koin, and serves as the central navigation hub for the application. (`appNavigation/build.gradle.kts`, `appNavigation/proguard-rules.pro`) [[1]](diffhunk://#diff-3fb6d216f2e523f6a5ca95763af12f080b89b8d4336ae2e88ae6783737300975R1-R61) [[2]](diffhunk://#diff-07f4dd09265d565e8fb1d514b1fc9955276b64321f19fa34437cee7d1860ac01R1-R21)

* Introduced the `FeatureAPIModule` to manage feature-specific navigation APIs for `Search` and `MediaDetails` features. (`app/src/main/java/com/paris_2/aflami/di/FeaturesApiModule.kt`)

### UI Updates:
* Removed the old theme files (`Color.kt`, `Theme.kt`, `Type.kt`) and transitioned to using a centralized design system module. (`app/src/main/java/com/paris_2/aflami/ui/theme/`) [[1]](diffhunk://#diff-c9f14d5e0d707c8fb0e18ba78a4a137c25c72b12af4d236d5d5465663b9c5638L1-L11) [[2]](diffhunk://#diff-1bccc4def50c18a66843bc869675885cefaaa81bcb504fd695ccb47454cf5d81L1-L58) [[3]](diffhunk://#diff-3561a48cee01bdce571cad7bf3ccbc61ee9e1df806dd5efb9afa610f433c2fc0L1-L34)

* Updated `MainActivity` to use Jetpack Compose for setting the app's content, integrating the new `AppNavGraph` and `AflamiTheme`. (`app/src/main/java/com/paris_2/aflami/MainActivity.kt`)

### Miscellaneous:
* Added a `.gitignore` file to the `appNavigation` module to exclude build artifacts. (`appNavigation/.gitignore`)

Joseph-Sameh-0 committed:
remove unnecessary build.gradle.kts

Joseph-Sameh-0 committed:
remove unnecessary build.gradle

Joseph-Sameh-0 committed:
Remove Koin DI modules from searchUi feature

The Koin dependency injection modules previously located within the `feature/search/searchUi` module have been removed. This includes modules for:

- DataSources (`DataSourceModule.kt`)
- Network (`NetworkModule.kt`)
- Repositories (`repositoryModule.kt`)
- Room (`RoomModule.kt`)
- UseCases (`UseCaseModule.kt`)
- ViewModels (`ViewModelModule.kt`)
- WorkManager (`WorkManagerModule.kt`)

Additionally, the `SearchApp.kt` custom Application class within the `searchUi` feature, which was responsible for initializing Koin, has also been removed.

The `SearchModule.kt` from `datasource/remote/search` has also been removed.

Dependencies related to these modules have been adjusted in `app/build.gradle.kts` and `feature/search/searchUi/build.gradle.kts`. The `app` module now directly includes dependencies for `koin-workmanager`, `work-runtime-ktx`, and Room, along with project dependencies for the search-related data and domain layers. The `searchUi` feature's manifest no longer references `SearchApp`.

The main application class is now `AflamiApplication` as declared in `app/src/main/AndroidManifest.xml`.

Joseph-Sameh-0 committed:
Add new module for app navigation

This commit introduces a new Android library module named `appNavigation`.

The module includes:
- Basic AndroidManifest.xml
- Proguard rules (proguard-rules.pro and consumer-rules.pro)
- .gitignore file
- build.gradle.kts with necessary configurations and dependencies
- Example instrumented and unit tests

The module has also been added to `settings.gradle.kts` and `buildSrc/src/main/kotlin/Modules.kt`.

Joseph-Sameh-0 committed:
refactor: Move navigation components to appNavigation module

This commit moves navigation-related classes and interfaces from the `feature/search/searchUi` module to a new `appNavigation` module. This is done to centralize navigation logic and make it reusable across different features.

The following changes are included:

- Moved `Destination.kt`, `Graph.kt`, `NavigationEvent.kt`, `Navigator.kt`, `NavigatorImpl.kt`, and `ObserveAsEvents.kt` to the `appNavigation` module.
- Created `AppDestinations.kt` in `appNavigation` to define app-level navigation destinations and graphs.
- Created `AppNavGraph.kt` in `appNavigation` to define the main application navigation graph.
- Updated `SearchNavGraph.kt` in `feature/search/searchUi` to use the new `appNavigation` components and define its own graph structure.
- Renamed `Destinations.kt` to `SearchDestinations.kt` within the `searchUi` module.
- Updated `ViewModelModule.kt` in the `app` module to use the new `Navigator` from `appNavigation`.
- Updated ViewModels in `searchUi` to use `SearchDestinations` and `Navigator` from `appNavigation`.
- Added `appNavigation` module dependency to `feature/search/searchUi` and `app` modules.
- Removed Koin dependencies from `feature/search/searchUi/build.gradle.kts` as they are now handled by `appNavigation`.
- Enabled Compose and buildConfig in `appNavigation/build.gradle.kts`.
- Added navigation-compose and Koin dependencies to `appNavigation/build.gradle.kts`.

Joseph-Sameh-0 committed:
refactor navigation and modularize search feature

Replaces legacy navigation interfaces with new AppNavigator and SearchNavigator abstractions, updating all usages and event handling accordingly. Moves search navigation contracts to a new feature/search/searchApi module, introduces SearchFeatureAPI for modular feature integration, and updates dependency injection modules. Removes obsolete theme and MainActivity files, and updates build scripts to support the new module structure and serialization. This enables a more scalable, modular navigation architecture and decouples the search feature implementation.

Joseph-Sameh-0 committed:
Rename Graph.kt to SearchGraph.kt

Joseph-Sameh-0 committed:
Implement MediaDetails feature API and the navigation

This commit introduces the MediaDetails feature module, including its API, UI, navigation, and ViewModel.

**Key changes:**

- **MediaDetails API (`feature/mediaDetails/mediaDetailsApi`):**
    - Defined `MediaDetailsDestination`, `MediaDetailsGraph`, and `MediaDetailsDestinations` for navigation within the feature.
    - Added serialization/deserialization functions (`toJson`, `fromJsonToMediaDetailsDestination`).
    - Introduced `MediaDetailsFeatureAPI` interface for feature integration.
    - Updated `build.gradle.kts` to include necessary dependencies (Android library, Kotlin, Compose, Serialization).
- **MediaDetails UI (`feature/mediaDetails/mediaDetailsUi`):**
    - Implemented `MediaDetailsFeatureAPIImpl` to provide the feature's Composable entry point.
    - Created `MediaDetailsViewModel` and `MediaDetailsScreen`.
    - Set up navigation components: `MediaDetailsNavGraph`, `MediaDetailsNavigator`, `MediaDetailsNavigatorImpl`, `MediaDetailsNavigationEvent`, and `ObserveAsEvents`.
    - Added `BaseViewModel` for common ViewModel logic.
    - Removed obsolete `MainActivity.kt` and theme files (`Color.kt`, `Theme.kt`, `Type.kt`).
    - Updated `build.gradle.kts` with dependencies for Navigation, Koin, MediaDetails API, and DesignSystem.
- **App Navigation (`appNavigation`):**
    - Updated `AppNavGraph.kt` (renamed from `SearchNavGraph.kt`) to include navigation to the `MediaDetailsFeature`.
    - Modified `AppDestinations.kt` to define `MediaDetailsFeature` destination.
    - Added `FEATURE_MEDIA_DETAILS_API` dependency to `build.gradle.kts`.
- **Search Feature (`feature/search/searchUi`, `feature/search/searchApi`):**
    - Updated ViewModels (`FindByActorViewModel`, `SearchViewModel`, `WorldTourViewModel`) to use `AppNavigator` for navigating to the `MediaDetailsFeature`.
    - Modified `SearchDestination.kt` to use `sealed interface` and updated serialization/deserialization functions.
    - Added `FEATURE_MEDIA_DETAILS_API` dependency to `build.gradle.kts`.
- **App Module (`app`):**
    - Updated `FeaturesApiModule.kt` to include Koin bindings for `MediaDetailsNavigator` and `MediaDetailsFeatureAPI`.
    - Updated `ViewModelModule.kt` to include `MediaDetailsViewModelViewModel`.
    - Added `FEATURE_MEDIA_DETAILS_API` and `FEATURE_MEDIA_DETAILS_UI` dependencies to `build.gradle.kts`.

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/refactor-navigation

Joseph-Sameh-0 committed:
Update release notes to version 0.1.18

Joseph-Sameh-0 committed:
refactor: Separate movie and TV show details screens

This commit introduces separate screens and ViewModels for movie and TV show details, enhancing code organization and maintainability.

Key changes include:

-   **MediaDetailsDestinations:**
    -   Renamed `MediaDetailsScreen` to `MovieDetailsScreen` and changed its parameter from `mediaId` to `movieId`.
    -   Added a new destination `TvShowDetailsScreen` with a `tvShowId` parameter.
-   **MediaDetailsViewModel:**
    -   Renamed `MediaDetailsViewModel` to `MovieDetailsViewModelViewModel`.
    -   Updated state data class to `MovieDetailsViewModelScreenState`.
    -   Modified initialization to use `movieId` from `MovieDetailsScreen` destination.
    -   Navigation logic updated to navigate to `TvShowDetailsScreen`.
    -   Created a new `TvShowDetailsViewModelViewModel` and `MediaDetailsViewModelScreenState` for TV show details.
-   **SearchResultContent:**
    -   Modified `onMediaCardClick` to accept both `id` and `MediaTypeUi`.
-   **Search Feature ViewModels (FindByActorViewModel, WorldTourViewModel, SearchViewModel):**
    -   Updated `onMediaCardClick` to accept `mediaType` parameter.
    -   Modified navigation logic to conditionally navigate to either `MovieDetailsScreen` or `TvShowDetailsScreen` based on `mediaType`.
-   **MediaDetailsNavGraph:**
    -   Updated navigation graph to include composables for both `MovieDetailsScreen` and `TvShowDetailsScreen`.
    -   Default start destination changed to `MovieDetailsScreen`.
-   **Screen Interaction Listeners:**
    -   Updated `onMediaCardClick` in `FindByActorScreenInteractionListener`, `SearchScreenInteractionListener`, and `WorldTourScreenInteractionListener` to include `mediaType`.
-   **UI Screens:**
    -   Renamed `MediaDetailsScreen` (the composable) to `TvShowDetailsScreen` and updated it to use `TvShowDetailsViewModelViewModel`.
    -   Created a new `MovieDetailsScreen` composable that uses `MovieDetailsViewModelViewModel`.
-   **Dependency Injection:**
    -   Updated `ViewModelModule` to provide `MovieDetailsViewModelViewModel` and `TvShowDetailsViewModelViewModel`.
    -   Commented out `FireBaseModule`.

Joseph-Sameh-0 committed:
Lower code coverage minimum to 20%

Joseph-Sameh-0 committed:
Temporarily set Kover minimum coverage to 0

mohammed-akkad commented on app/src/main/java/com/paris_2/aflami/di/FireBaseModule.kt:
```diff
@@ -1,9 +1,9 @@
-    single<Logger> { FireBaseCrashlyticsLogger() }
-}
+//import com.parise_2.firebase.firebase.FireBaseCrashlyticsLogger
+//import com.parise_2.firebase.repo.Logger
```
why this code is comment?

mohammed-akkad commented on appNavigation/build.gradle.kts:
```diff
@@ -0,0 +1,61 @@
+
+android {
+    namespace = "com.paris_2.aflami.appnavigation"
+    compileSdk = 35
```
use configration instead of hardcode

mohammed-akkad commented on feature/mediaDetails/mediaDetailsUi/src/main/java/com/feature/mediaDetails/mediaDetailsUi/ui/comon/BaseViewModel.kt:
```diff
@@ -0,0 +1,55 @@
+    private val privateScreenState = MutableStateFlow(initialState)
+    val screenState: StateFlow<S> = privateScreenState.asStateFlow()
+
+    private val navigator: MediaDetailsNavigator by KoinJavaComponent.inject(MediaDetailsNavigator::class.java)
```
why MediaDetailsNavigator in base view model?

Joseph-Sameh-0 replied:
Because this is the BaseViewModel of the Media details

Joseph-Sameh-0 committed:
refactor: Inject AppNavigator into ViewModels

Injects AppNavigator into `FindByActorViewModel`, `SearchViewModel`, and `WorldTourViewModel` through their constructors instead of retrieving it directly via Koin's `get()` method.

This change also updates the corresponding test files to mock the `AppNavigator` dependency.

Joseph-Sameh-0 committed:
Update compileSdk in appNavigation module

Joseph-Sameh-0 committed:
Initialize AppNavigator in SearchViewModelTest

This commit updates the SearchViewModelTest to properly initialize the `appNavigator` dependency. Previously, it was missing, causing potential issues in test execution.

Additionally, the `FireBaseModule` has been uncommented, enabling Firebase Crashlytics logging functionality.

Joseph-Sameh-0 committed:
Update Koin dependency injection in BaseViewModels

This commit updates the Koin dependency injection mechanism in `BaseViewModel.kt` for both the `searchUi` and `mediaDetailsUi` modules.

The `KoinJavaComponent.inject()` method has been replaced with the more idiomatic Kotlin `inject()` extension function provided by `org.koin.core.component.inject`.

This change simplifies the code and aligns it with common Kotlin Koin usage patterns.

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/refactor-navigation

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/refactor-navigation

Joseph-Sameh-0 committed:
Remove MainActivity declaration from mediaDetailsUi manifest

The `MainActivity` declaration has been removed from the `AndroidManifest.xml` file within the `feature/mediaDetails/mediaDetailsUi` module. This activity is no longer needed in this specific module.

Joseph-Sameh-0 committed:
Exclude navigation packages from Kover reports

This commit updates the `build.gradle.kts` file to exclude files within `**.navigation.**` packages from Kover code coverage reports. This ensures that navigation-related code, which may not be suitable for unit testing, does not affect the overall coverage metrics.

Joseph-Sameh-0 committed:
Lower code coverage minimum in searchUi module

The minimum code coverage requirement for the `feature/search/searchUi` module has been reduced from 75% to 70%.
